### ISSUE TITLE
Allow collection `while` 

### ISSUE THREAD

**Issue by Zekfad** [dart-lang/language#3093]
It's inconvenient to white `for (;cond;)` when you really mean `while(cond)` inside of collection.
I'd like to allow `while` to be used inside collection expansion, like that:
```dart
final objects = [
  while (reader.remainingBytes > 0)
    reader.readStruct(),
];
```
That would be juts a syntax sugar for `for (;;)`, but without need of executing variables initialization and post-loop action.
`do`-`while` on the otherhand is useful, but probably would make syntax bit more complex and ambiguous (in collection `do` what?).

**Comment by lrhn** [comment]
I can't see any syntactic reason this couldn't work. Good old `while (iterator.moveNext()) if (iterator.current.isEvent) iterator.current` would work.

The `do`/`while` could be harder to parse, since elements are not usually delimited. A `do element while (expression)` needs to be certain that the `while` is not starting another `while`-element.
It's probably doable, but can also get hard to read. Maybe if we require more parentheses: `do (element) while (expression)`.

**Comment by munificent** [comment]
> The `do`/`while` could be harder to parse, since elements are not usually delimited. A `do element while (expression)` needs to be certain that the `while` is not starting another `while`-element.

I don't think that would be a problem. For it to start another `while` element, there would have to be a comma after the preceding element.

**Comment by lrhn** [comment]
I'm not so much worried about sequential element parsing, as about nested element parsing.
```dart
var x = 1;
var l  = [do while (x > 0) while (x > 0) x-- while (x >0)];
```
I'm somewhat convinced that it's non-ambiguous, because there has to be a leaf-element somewhere, but I'm not *absolutely* sure there isn't some tricky exception I'm missing.
